exchanges/account/account.go

```diff
@@ -39,80 +52,255 @@ func CollectBalances(accountBalances map[string][]Balance, assetType asset.Item)
+	}
+
+	if a.Exchange == "" {
+		return errExchangeNameUnset
+	}
+
 	exch := strings.ToLower(a.Exchange)
-	s.Lock()
-	acc, ok := s.accounts[exch]
+	s.m.Lock()
+	defer s.m.Unlock()
+	accounts, ok := s.exchangeAccounts[exch]
 	if !ok {
 		id, err := s.mux.GetID()
 		if err != nil {
-			s.Unlock()
 			return err
 		}
+		accounts = &Accounts{
+			ID:          id,
+			SubAccounts: make(map[string]map[asset.Item]map[*currency.Item]*ProtectedBalance),
+		}
+		s.exchangeAccounts[exch] = accounts
+	}
+
+	var errs common.Errors
+	for x := range a.Accounts {
+		if !a.Accounts[x].AssetType.IsValid() {
+			errs = append(errs, fmt.Errorf("cannot load sub account holdings for %s [%s] %w",
+				a.Accounts[x].ID,
+				a.Accounts[x].AssetType,
+				asset.ErrNotSupported))
+			continue
+		}
+
+		lowerSA := strings.ToLower(a.Accounts[x].ID)
+		var accountAssets map[asset.Item]map[*currency.Item]*ProtectedBalance
+		accountAssets, ok = accounts.SubAccounts[lowerSA]
+		if !ok {
+			accountAssets = make(map[asset.Item]map[*currency.Item]*ProtectedBalance)
+			accounts.SubAccounts[lowerSA] = accountAssets
+		}
+
+		var currencyBalances map[*currency.Item]*ProtectedBalance
+		currencyBalances, ok = accountAssets[a.Accounts[x].AssetType]
+		if !ok {
+			currencyBalances = make(map[*currency.Item]*ProtectedBalance)
+			accountAssets[a.Accounts[x].AssetType] = currencyBalances
+		}
 
-		s.accounts[exch] = &Account{h: a, ID: id}
-		s.Unlock()
-		return nil
+		for y := range a.Accounts[x].Currencies {
+			bal := currencyBalances[a.Accounts[x].Currencies[y].CurrencyName.Item]
+			if bal == nil {
+				bal = &ProtectedBalance{}
+				currencyBalances[a.Accounts[x].Currencies[y].CurrencyName.Item] = bal
+			}
+			bal.load(a.Accounts[x].Currencies[y])
+		}
+	}
+	err := s.mux.Publish([]uuid.UUID{accounts.ID}, a)
+	if err != nil {
+		return err
+	}
+
+	if errs != nil {
+		return errs
+	}
+
+	return nil
+}
+
+// load checks to see if there is a change from incoming balance, if there is a
+// change it will change then alert external routines.
+func (b *ProtectedBalance) load(change Balance) {
+	b.m.Lock()
+	defer b.m.Unlock()
+	if b.total == change.Total &&
+		b.hold == change.Hold &&
+		b.free == change.Free &&
+		b.availableWithoutBorrow == change.AvailableWithoutBorrow &&
+		b.borrowed == change.Borrowed {
+		return
 	}
+	b.total = change.Total
+	b.hold = change.Hold
+	b.free = change.Free
+	b.availableWithoutBorrow = change.AvailableWithoutBorrow
+	b.borrowed = change.Borrowed
+	b.notice.Alert()
+}
+
+// Wait waits for a change in amounts for an asset type. This will pause
+// indefinitely if no change ever occurs. Max wait will return true if it failed
+// to achieve a state change in the time specified. If Max wait is not specified
+// it will default to a minute wait time.
+func (b *ProtectedBalance) Wait(maxWait time.Duration) (wait <-chan bool, cancel chan<- struct{}, err error) {
+	if b == nil {
+		return nil, nil, errBalanceIsNil
+	}
+
+	if maxWait == 0 {
```

Comment:
```suggestion
	if maxWait <= 0 {
```
To counter any scallywags out there

Reply:
SCALLYWAGS EVERYWHERE!